File: python/Permutations.py
class Solution(object):
    answer = None
    used = None
    n = None
    nums = None
    def dfs(self, x, _tmp):
        tmp = copy.copy(_tmp)
        tmp.append(x)
        if len(tmp) == self.n:
            self.answer.append(tmp)
            return
        for num in self.nums:
            if not self.used[num]:
                self.used[num] = True
                self.dfs(num, tmp)
                self.used[num] = False
        
    def permute(self, nums):
        self.used = dict()
        self.nums = nums
        for num in nums:
            self.used[num] = False

        self.n = len(nums)
        self.answer = []
        for num in nums:
            self.used[num] = True
            self.dfs(num, [])
            self.used[num] = False
        return self.answer

File: cpp/Permutations.cpp
class Solution {
public:
    bool used[1000];
    vector<int> num;
    int n;
    vector<vector<int>> answer;
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        num = nums;
        for (int i = 0; i < n; i++) {
            used[i] = false;
        }
        
        for (int i = 0; i < n; i++) {
            if (!used[i]) {
                vector<int> list;
                list.push_back(i);
                used[i] = true;
                helper(i, list);
                list.pop_back();
                used[i] = false;
            }
        }
        return answer;
    }
    
    void helper(int x, vector<int> &list) {
        if (list.size() == n) {
            vector<int> tmp;
            for (int i = 0; i < n; i++) {
                tmp.push_back(num[list[i]]);
            }
            answer.push_back(tmp);
            return;
        }
        for (int i = 0; i < n; i++) {
            if (!used[i]) {
                used[i] = true;
                list.push_back(i);
                helper(i, list);
                list.pop_back();
                used[i] = false;
            }
        }
    }
};
